openmoc/materialize.py

```diff
@@ -355,3 +370,347 @@ def load_openmc_mgxs_lib(mgxs_lib, geometry=None):
+            continue
+
+        # Compute the total volume filled by this domain throughout the geometry
+        tot_volume = openmoc_domain.getVolume()
+
+        # Extract an openmc.mgxs.MGXS object for the scattering matrix
+        if 'nu-scatter matrix' in mgxs_lib.mgxs_types:
+            scatter = mgxs_lib.get_mgxs(openmoc_domain.getId(),
+                                        'nu-scatter matrix')
+        elif 'scatter matrix' in mgxs_lib.mgxs_types:
+            scatter = mgxs_lib.get_mgxs(openmoc_domain.getId(),
+                                        'scatter matrix')
+        else:
+            py_printf('ERROR', 'Unable to compute SPH factors for an OpenMC '
+                               'MGXS library without scattering matrices')
+
+        # Extract an openmc.mgxs.MGXS object for the nu-fission cross section
+        if 'nu-fission' in mgxs_lib.mgxs_types:
+            nu_fission = mgxs_lib.get_mgxs(openmoc_domain.getId(), 'nu-fission')
+        else:
+            py_printf('ERROR', 'Unable to compute SPH factors for an OpenMC '
+                               'MGXS library without nu-fission cross sections')
+
+        # Extract an openmc.mgxs.MGXS object for the chi fission spectrum
+        if 'chi' in mgxs_lib.mgxs_types:
+            chi = mgxs_lib.get_mgxs(openmoc_domain.getId(), 'chi')
+        else:
+            py_printf('ERROR', 'Unable to compute SPH factors for an OpenMC '
+                               'MGXS library without chi fission spectrum')
+
+        # Retrieve the OpenMC volume-integrated flux for this domain from
+        # the nu-fission MGXS and store it for SPH factor calculation
+        flux = nu_fission.tallies['flux'].mean.flatten()
+        openmc_fluxes[i, :] = np.atleast_1d(np.flipud(flux))
+        openmc_fluxes[i, :] /= tot_volume
+
+        # Extract a NumPy array for each MGXS summed across all nuclides
+        scatter = scatter.get_xs(nuclides='sum')
+        nu_fission = nu_fission.get_xs(nuclides='sum')
+        chi = chi.get_xs(nuclides='sum')
+
+        # Compute and store volume-averaged fission + scatter sources
+        for group in range(num_groups):
+
+            # Compute the source for this group from fission and scattering
+            in_scatter = scatter[:, group] * openmc_fluxes[i, :]
+            fission = (chi[group] / keff) * nu_fission * openmc_fluxes[i, :]
+            source = np.sum(in_scatter) + np.sum(fission)
+
+            # Assign the source to this domain
+            fsr_ids = domains_to_fsrs[openmoc_domain.getId()]
+            for fsr_id in fsr_ids:
+                solver.setFixedSourceByFSR(fsr_id, group+1, source)
+
+    return openmc_fluxes
+
+
+##
+# @brief Apply SPH factors to an OpenMC MGXS library
+# @details This is a helper routine for openmoc.process.compute_sph_factors(...)
+# @param mgxs_lib an openmc.mgxs.Library object
+# @param sph a NumpPy array of SPH factors for each energy group for this domain
```

Comment:
Great point, we certainly want the comments to be in the right order (esp. if we switch to docstring comments as I think we should).